rustworkx-core/src/coloring.rs

```diff
@@ -220,8 +221,218 @@ where
     edge_colors
 }
 
-#[cfg(test)]
+struct MisraGries<G: GraphBase> {
+    // The input graph
+    graph: G,
+    // Partially assigned colors (indexed by internal edge index)
+    colors: Vec<Option<usize>>,
+}
+
+impl<G> MisraGries<G>
+where
+    G: EdgeIndexable + IntoEdges,
+{
+    pub fn new(graph: G) -> Self {
+        let colors = vec![None; graph.edge_bound()];
+        MisraGries { graph, colors }
+    }
+
+    // Computes colors used at node u
+    fn get_used_colors(&self, u: G::NodeId) -> HashSet<usize> {
+        let used_colors: HashSet<usize> = self
+            .graph
+            .edges(u)
+            .filter_map(|edge| self.colors[EdgeIndexable::to_index(&self.graph, edge.id())])
+            .collect();
+        used_colors
+    }
+
+    // Returns the smallest free (aka unused) color at node u
+    fn get_free_color(&self, u: G::NodeId) -> usize {
+        let used_colors = self.get_used_colors(u);
+        let free_color: usize = (0..)
```

Comment:
Just to avoid unbound iteration as a potential source of bugs:

```suggestion
        let free_color: usize = (0..self.colors.len())
```

Reply:
Fixed in [4e22ad7](https://github.com/Qiskit/rustworkx/pull/902/commits/4e22ad7abdb6c27eac28c0ea88927d94ea50eb93) using the stricter upper bound of `max_node_degree + 1`.